test/libsolidity/smtCheckerTests/overflow/safe_mul_2.sol

```diff
@@ -0,0 +1,15 @@
+pragma experimental SMTChecker;
+
+contract C
+{
+	function mul(uint256 a, uint256 b) internal pure returns (uint256) {
+		if (a == 0) {
+			return 0;
+		}
+		uint256 c = a * b;
+		require(c / a == b);
+		return c;
+	}
+}
+// ----
+// Warning: (180-185): Division by zero happens here
```

Comment:
This false positive happens because the SMTChecker does not see the code after the `if` as an else, which it should due to the early return.

Reply:
Can we design an internal representation where instead of adding certain conditions (such as early return) we can simplify the AST? In this example wrap the other side into an else block.

Would such an option be worse complexity wise?

Reply:
We actually wanted to do that to simplify some operations, for example compound assignments. Including that is a good idea.

Reply:
I agree that we should do that for certain cases, but, here I would actually say that a reachability analysis should be done instead, because we should do it anyway. Think of, for example
```
if (a == 0) {
  if (anotherComplicatedCondition()) return 0;
}
```
If we can prove that `anotherComplicatedCondition` always results in true, AST rewriting will not help, but reachability analysis will, and we should already have all the facts needed for reachability analysis, it's just the path constraints.

Currently, we check dead code by checking the condition of an if, perhaps we should instead check the combined path conditions after the if.

Reply:
What does "after the if" mean? We currently check the conjunction of path conditions with the if condition.